### DISCUSSION TITLE
Xcode 16 - SnapshotTesting in a Swift Package

### DISCUSSION THREAD

**Discussion by kelvinlauKL** [pointfreeco/swift-snapshot-testing#1015] [General]
I'm building a UI library as a Swift Package. Attempting to use `SnapshotTesting` gives me this:

<img width="1520" alt="Screenshot 2024-07-28 at 8 38 23 PM" src="https://github.com/user-attachments/assets/de652a2c-bb18-4d96-866f-21740654df47">

Here's my package file:

```
// swift-tools-version: 6.0
// The swift-tools-version declares the minimum version of Swift required to build this package.

import PackageDescription

let package = Package(
  name: "ChatDesignSystem",
  platforms: [.macOS(.v15), .iOS(.v18), .tvOS(.v18), .watchOS(.v11), .macCatalyst(.v18)],
  products: [
    .library(
      name: "ChatDesignSystem",
      targets: ["ChatDesignSystem"]),
  ],
  dependencies: [
    .package(url: "https://github.com/pointfreeco/swift-snapshot-testing", branch: "main")
  ],
  targets: [
    .target(
      name: "ChatDesignSystem"),
    .testTarget(
      name: "ChatDesignSystemTests",
      dependencies: [
        "ChatDesignSystem",
        .product(name: "SnapshotTesting", package: "swift-snapshot-testing")
      ]
    ),
  ]
)

```

Is this not supported in a Swift Package?

**Comment by stephencelis** [comment]
@kelvinlauKL The screenshot has the error message mostly truncated. It looks like a compile failure but it's not easy to troubleshoot as is. Can you share sample code reproducing the issue?

**Comment by kelvinlauKL** [comment]
My bad - the error message is `Type of expression is ambiguous without a type annotation`. I realized this compiler error is happening when my device is set to "My Mac". The compiler error goes away if I set it to an iOS device.

I've updated the screenshot.

Is that expected behaviour?

**Comment by doozMen** [comment]
I have the same issue and was looking into it. I found out to my surprise that the current library does not seam to have the image not compiled for MacOS as the current snapshot implementation seams to rely on making the view a UIImage and embedding it in a viewController. Both do not exist on macOS.

So SwiftUI view support is limited to iOS like platforms
<img width="848" alt="Screenshot 2024-08-05 at 14 45 20" src="https://github.com/user-attachments/assets/719d7524-93d6-495b-895b-2bb9bbe5c512">

**Comment by gsabran** [comment]
Bummer! It would be great to have similar support for MacOS

**Comment by iSapozhnik** [comment]
Hey @stephencelis! Is there a reason it's not available for mac? I've been working on the design system for my mac apps and decided to use snapshotting to verify and showcase implemented components and hit the same problem as other guys from the thread.

**Comment by mbrandonw** [comment]
@iSapozhnik, @gsabran, @doozMen: The library is extensible so that you can add your own snapshot strategies if the ones in the library do not suite your needs. You will see that in the screenshot above, the strategy is defined as:

```swift
extension Snapshotting where Value: SwiftUI.View, Format == UIImage {
  …
}
```

So, you will want to try to make an equivalent for `NSImage`:

```swift
extension Snapshotting where Value: SwiftUI.View, Format == NSImage {
  …
}
```

Give that a shot and let us know how it goes.

This is not considered an issue with the library currently, and so I am going to convert this to a discussion. Feel free to keep the conversation going over there.

**Reply by iSapozhnik** [reply]
So far I've been converting it to `NSView` like so:

```swift
fileprivate enum SnapshotUtility<Content> where Content: View {

    static func view(from content: Content, colorScheme: ColorScheme) -> NSView {
        let nsView = NSHostingView(
            rootView: content
                .environment(\.colorScheme, colorScheme)
                .fixedSize()
        )
        nsView.layoutSubtreeIfNeeded()
        nsView.frame = NSRect(origin: .zero, size: nsView.fittingSize)
        return nsView
    }
}

```
and although it works, it's ugly :) Let me try your approach.

**Reply by anuragajwani** [reply]
is there a plan to bring SwiftUI MacOS snapshotting strategies to the library?
